drivers/gpio/gpio_gd32.c

```diff
@@ -0,0 +1,363 @@
+					 gpio_port_pins_t mask,
+					 gpio_port_value_t value)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	GPIO_OCTL(config->reg) =
+		(GPIO_OCTL(config->reg) & ~mask) | (value & mask);
+
+	return 0;
+}
+
+static int gpio_gd32_port_set_bits_raw(const struct device *port,
+				       gpio_port_pins_t pins)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	GPIO_BOP(config->reg) = pins;
+
+	return 0;
+}
+
+static int gpio_gd32_port_clear_bits_raw(const struct device *port,
+					 gpio_port_pins_t pins)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	GPIO_BC(config->reg) = pins;
+
+	return 0;
+}
+
+static int gpio_gd32_port_toggle_bits(const struct device *port,
+				      gpio_port_pins_t pins)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+#ifdef CONFIG_GD32_HAS_AF_PINMUX
+	GPIO_TG(config->reg) = pins;
+#else
+	GPIO_OCTL(config->reg) ^= pins;
+#endif /* CONFIG_GD32_HAS_AF_PINMUX */
+
+	return 0;
+}
+
+static int gpio_gd32_pin_interrupt_configure(const struct device *port,
+					     gpio_pin_t pin,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
+{
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		gd32_exti_disable(pin);
+		(void)gd32_exti_configure(pin, NULL, NULL);
+		gd32_exti_trigger(pin, GD32_EXTI_TRIG_NONE);
+	} else if (mode == GPIO_INT_MODE_EDGE) {
+		int ret;
+
+		ret = gpio_gd32_configure_extiss(port, pin);
+		if (ret < 0) {
+			return ret;
+		}
+
+		switch (trig) {
+		case GPIO_INT_TRIG_LOW:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_FALLING);
+			break;
+		case GPIO_INT_TRIG_HIGH:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_RISING);
+			break;
+		case GPIO_INT_TRIG_BOTH:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_BOTH);
+			break;
+		default:
+			return -ENOTSUP;
+		}
+
+		ret = gd32_exti_configure(pin, gpio_gd32_isr, (void *)port);
+		if (ret < 0) {
+			return ret;
+		}
+
+		gd32_exti_enable(pin);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int gpio_gd32_manage_callback(const struct device *dev,
+				     struct gpio_callback *callback, bool set)
+{
+	struct gpio_gd32_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static const struct gpio_driver_api gpio_gd32_api = {
+	.pin_configure = gpio_gd32_configure,
+	.port_get_raw = gpio_gd32_port_get_raw,
+	.port_set_masked_raw = gpio_gd32_port_set_masked_raw,
+	.port_set_bits_raw = gpio_gd32_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_gd32_port_clear_bits_raw,
+	.port_toggle_bits = gpio_gd32_port_toggle_bits,
+	.pin_interrupt_configure = gpio_gd32_pin_interrupt_configure,
+	.manage_callback = gpio_gd32_manage_callback,
+};
+
+static int gpio_gd32_init(const struct device *port)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	rcu_periph_clock_enable(config->rcu_periph_clock);
+
+#ifdef CONFIG_GD32_HAS_AF_PINMUX
+	/* enable access to SYSCFG_EXTISSn registers */
+	rcu_periph_clock_enable(RCU_SYSCFG);
```

Comment:
For AF_PINMUX, RCU_SYSCFG are not always exist, GD32F3x0 doesn't have this clock control bit. 
Not sure which way is more general, just leave a comment in here.

Reply:
solved by using DT (check syscfg node)